feat: adds 'How to Migrate off Terraform Cloud' post (#28)
masterpointio/masterpoint.io: content/updates/migrate-off-tfc.md

---
visible: true
draft: false
title: "How to Migrate off Terraform Cloud"
author: Veronika Gnilitska
slug: how-to-migrate-off-tfc
date: 2024-07-25
description: "Need to to migrate off Terraform Cloud? We're happy to share some tips about preparation, pitfalls, and the process itself based on Masterpoint's experience."
image: /img/updates/migrate-off-tfc.jpeg
callout: <p>👋 <b>If you're ready to take your infrastructure to the next level, we're here to help. We love to work together with engineering teams to help them build well-documented, scalable, automated IaC that make their jobs easier. <a href='/contact'>Get in touch!</a>
---

# Introduction

> Migrating off Terraform Cloud could seem a dreaded task, but with proper planning and execution, it can be done smoothly and efficiently. Whether you're moving due to cost considerations, the need for more control, or other reasons, this post will walk you through the essential steps and action items to successfully transition your infrastructure management away from Terraform Cloud.

Reviewer: I suggest we shorted to TFC. Thoughts?

```suggestion
Migrating off Terraform Cloud (TFC) could seem a dreaded task, but with proper planning and execution, it can be done smoothly and efficiently. Whether you're moving due to cost considerations, the need for more control, or licensing considerations, this post will walk you through the essential steps and action items to successfully transition your infrastructure management away from TFC.
```

Author: Hm, considering [this](https://developer.hashicorp.com/terraform/cloud-docs):
>Effective April 22, 2024, Terraform Cloud is now HCP Terraform. 

Should we be consistent with the updated naming and change Terraform Cloud to HCP Terraform?

Author: `licensing considerations` suggestion added to the doc

> 

Reviewer: ```suggestion

As a quick aside, we have nothing against TFC or Hashicorp. The product is full featured, does a great job at automating Terraform, and the team building it has all of our respect. We do know multiple clients, prospects, and colleagues within the industry who are making the switch away due to their pricing model, the recent acquisition by IBM, or the change in licensing that occurred last year. We want to support that group within our community and share our experience, which is the goal of this blog post. 
```

Author: Added to the doc 👍

As a quick aside, we have nothing against HCP Terraform or HashiCorp. The product is full-featured, does a great job at automating Terraform, and the team building it has our respect. We do know many clients, prospects, and colleagues within the industry who are switching away due to their pricing model, the recent acquisition by IBM, or the change in licensing that occurred last year.

The goal of this post is to support that portion of the community as well as share our experience.

At the end of this post, you’ll learn how to migrate away from Terraform Cloud successfully.

# Preparation for Migration

Before diving into the actual migration process, it's crucial to assess your current usage of Terraform Cloud. Here is what you need to do:

1. Decide on the platform to migrate to.
1. Locate your state files. You will need to migrate these state files if they're stored in Terraform Cloud.
1. Evaluate the variables and secrets stored within Terraform Cloud.
1. Evaluate any additional Terraform Cloud use cases or tools that you rely on (for example, drift detection, Infracost, Sentinel, or similar tools).

> 1. Consider and define any new or different  configuration your IaC migration destination requires.

Reviewer: ```suggestion
1. Consider and define any new or different configuration that your new IaC migration destination requires.
```

It’s important to note that simply changing the backend configuration (e.g., from Terraform Cloud to S3) when running `terraform init` will **not** automatically migrate your state like it will for other state backends. Instead, you’ll encounter an error, as Terraform does not yet support automated migration from TFC:

```sh
❯ terraform init

Initializing the backend...
Migrating from Terraform Cloud to backend "s3".
╷
│ Error: Migrating state from Terraform Cloud to another backend is not yet implemented.
│
│ Please use the API to do this: https://www.terraform.io/docs/cloud/api/state-versions.html

```

To migrate the state, you must use the Terraform [API](https://www.terraform.io/docs/cloud/api/state-versions.html) to download your state files, which you can then upload to your preferred storage solution.

This extra step makes the migration process more labor-intensive compared to other backends. We’ve found [this detailed guide on downloading your Terraform state files to be helpful](https://github.com/hashicorp/terraform/issues/33214#issuecomment-1553223031). It is a GitHub comment by a HashiCorp team member.

> Ensure your new storage solution is configured correctly to handle Terraform state files, including appropriate redundancy, access permissions, and encryption settings. While there are other options, we encourage our clients to utilize their primary cloud’s object storage solution, like AWS S3, Google Cloud Storage, or Azure’s Blob Storage, to ensure that the infrastructure management process is secure, reliable, and integrated within the broader cloud environment. If you want a jumpstart on using S3 as your state storage, check out Cloud Posse’s tfstate-backend module.

Reviewer: ```suggestion
Ensure your new storage solution is configured correctly to handle Terraform state files, including appropriate redundancy, access permissions, and encryption settings. While there are other options, we encourage our clients to utilize their primary cloud’s object storage solution, like AWS S3, Google Cloud Storage, or Azure’s Blob Storage, to ensure that the infrastructure management process is secure, reliable, and integrated within the broader cloud environment. If you want a jumpstart on using S3 as your state storage, [check out Cloud Posse’s tfstate-backend module](https://github.com/cloudposse/terraform-aws-tfstate-backend).
```

1. Do you run Drift Detection for any of your Terraform Cloud Workspaces? Do you need your new solution to have analogous functionality?

Finding the answers to these questions for your organization and coming up with a plan to deal with each use case in the planning phase will help ensure a smooth migration.

If you don’t use these tools but want to, plan for an after-migration project. It’s better to migrate functionality between Terraform Cloud and your new solution without improvements or unneeded changes to minimize migration complexity.

## Additional Destination Configuration

The destination you choose might have additional features, configuration, or functionality.

For example, if you’re going to migrate to Spacelift, there is configuration that either doesn’t have a direct equivalent in TFC or wasn’t available because of the tier level:
Do you want to tailor user permissions or workflow very specifically? Define your custom rules with [Rego-based](https://www.openpolicyagent.org/docs/latest/policy-language/) policies.

> Are you trying to solve the problem of managing segmentation and access control in large organizations or complex projects with multiple teams? Consider using [Spaces](https://docs.spacelift.io/concepts/spaces) feature.

Reviewer: ```suggestion
Are you trying to solve the problem of managing segmentation and access control in large organizations or complex projects with multiple teams? Consider using [Spaces](https://docs.spacelift.io/concepts/spaces) to section up your various Stacks.
```

- **Communication Plan**: Keep everyone informed about the migration progress and any potential disruptions.
- **Infrastructure Code Freeze**: Execute an infrastructure code freeze during the migration to prevent any unintended changes or conflicts.
- **Infrastructure Drift**: If you’ve been running Terraform operations against the infrastructure regularly, this is unlikely to happen. Be vigilant about infrastructure drift during the migration process and have a strategy to address it. This may require:
  - Updating variables to match existing resources.
  - Dealing with Terraform providers' updates.
  - Reviewing outstanding proposed changes and applying them.
- **Security Considerations**: Ensure that new storage solutions for state files and secrets comply with your organization's security policies.
- **Rollback Plan**: Have a rollback plan in place in case the migration encounters significant issues. This varies wildly based on your infrastructure, but at the minimum, you should prepare scripts that revert state migration and be ready to revert any changes within the infrastructure code.

## Checklist

> Here’s the promised [checklist](https://docs.google.com/document/d/1ibwIi3gKIx7KmhsnQtz8Xseb_gpLasC4wPMRyG57jiA/edit?usp=sharing). This list covers common tasks, but you should review and customize them for your migration project’s unique needs.

Reviewer: ```suggestion
Here’s the promised [checklist](https://docs.google.com/document/d/1ibwIi3gKIx7KmhsnQtz8Xseb_gpLasC4wPMRyG57jiA/edit?usp=sharing) in GDoc form, which you can easily copy. This list covers common tasks, but you should review and customize them for your migration project’s unique needs.
```
